flutter_kernel_transformers/lib/track_widget_constructor_locations.dart

```diff
@@ -0,0 +1,516 @@
+          importUri.scheme == 'package') {
+        if (importUri.path == 'flutter/src/widgets/framework.dart') {
+          for (Class class_ in library.classes) {
+            if (class_.name == 'Widget') {
+              _widgetClass = class_;
+            }
+          }
+        } else {
+          if (importUri.path == 'flutter/src/widgets/widget_inspector.dart') {
+            for (Class class_ in library.classes) {
+              if (class_.name == '_HasCreationLocation') {
+                _hasCreationLocationClass = class_;
+              } else if (class_.name == '_Location') {
+                _locationClass = class_;
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /// Modify [clazz] to add a field named [_locationFieldName] that is the
+  /// first parameter of all constructors of the class.
+  ///
+  /// This method should only be called for classes that implement but do not
+  /// extend [Widget].
+  void _transformClassImplementingWidget(Class clazz) {
+    if (clazz.fields
+        .any((Field field) => field.name.name == _locationFieldName)) {
+      // This class has already been transformed. Skip
+      return;
+    }
+    clazz.implementedTypes.add(new Supertype(_hasCreationLocationClass, <DartType>[]));
+    // We intentionally use the library context of the Widget class for the
+    // private field even if [clazz] is in a different library so that classes
+    // implementing Widget defined in other libraries behave just like
+    final Field locationField = new Field(
+        new Name(_locationFieldName, _hasCreationLocationClass.enclosingLibrary),
+        isFinal: true);
+    clazz.addMember(locationField);
+
+    final Set<Constructor> _handledConstructors =
+        new Set<Constructor>.identity();
+
+    void handleConstructor(Constructor constructor) {
+      if (!_handledConstructors.add(constructor)) {
+        return;
+      }
+      assert(!_hasNamedParameter(
+          constructor.function, _creationLocationParameterName));
+      final VariableDeclaration variable = new VariableDeclaration(
+          _creationLocationParameterName,
+          type: _locationClass.thisType);
+      if (!_maybeAddNamedParameter(constructor.function, variable)) {
+        return;
+      }
+
+      bool hasRedirectingInitializer = false;
+      for (Initializer initializer in constructor.initializers) {
+        if (initializer is RedirectingInitializer) {
+          if (initializer.target.enclosingClass == clazz) {
+            // We need to handle this constructor first or the call to
+            // addDebugLocationArgument bellow will fail due to the named
+            // parameter not yet existing on the constructor.
+            handleConstructor(initializer.target);
+          }
+          _maybeAddCreationLocationArgument(
+              initializer.arguments,
+              initializer.target.function,
+              new VariableGet(variable),
+              _locationClass);
```

Comment:
ditto

Reply:
there are others in this file. Basically any time dartfmt ends up doing weird things where you can't tell the nesting at a glance from the indentation, it's a good place to try adding commas until it does a better job.

Reply:
done

Reply:
done